File: swift/Sources/RiffCore/Draft.swift
import Foundation

public struct Line: Codable, Sendable {
    public struct Grounding: Codable, Sendable {
        public var ratio: Double
        public var kind: GroundingKind
        public var unmatchedTokens: [String]?
    }

    public var id: String
    public var section: Section
    public var text: String
    public var order: Double
    public var sourceUtteranceIds: [String]
    public var motifId: String?
    public var supersedes: [String]?
    public var grounding: Grounding
}

public struct ContextItem: Codable, Sendable {
    public var referenceId: String
    public var kind: String
    public var title: String
    public var identifier: String?
    public var url: String?
    public var actor: String?
    public var timestamp: String?
    public var state: String?
    public var summary: String?
    /// The phrase the speaker used, so the prompt and the resolved thing stay connected.
    public var resolvedFrom: String?
    public var confidence: Double?

    public init(
        referenceId: String,
        kind: String,
        title: String,
        identifier: String? = nil,
        url: String? = nil,
        actor: String? = nil,
        timestamp: String? = nil,
        state: String? = nil,
        summary: String? = nil,
        resolvedFrom: String? = nil,
        confidence: Double? = nil
    ) {
        self.referenceId = referenceId
        self.kind = kind
        self.title = title
        self.identifier = identifier
        self.url = url
        self.actor = actor
        self.timestamp = timestamp
        self.state = state
        self.summary = summary
        self.resolvedFrom = resolvedFrom
        self.confidence = confidence
    }
}

public struct Motif: Codable, Sendable {
    public var id: String
    public var text: String
    public var scope: String
    public var appliesWhen: String?
    public var createdAt: String
    public var retiredAt: String?

    public init(id: String, text: String, scope: String, appliesWhen: String? = nil, createdAt: String, retiredAt: String? = nil) {
        self.id = id
        self.text = text
        self.scope = scope
        self.appliesWhen = appliesWhen
        self.createdAt = createdAt
        self.retiredAt = retiredAt
    }
}

public enum TakeStatus: String, Codable, Sendable {
    case drafting, ready, submitted, parked, discarded
}

public struct DraftOperation: Sendable {
    public enum Kind: String, Sendable {
        case setTitle = "set_title"
        case upsertLine = "upsert_line"
        case removeLine = "remove_line"
        case moveLine = "move_line"
        case attachContext = "attach_context"
        case detachContext = "detach_context"
    }

    public var op: Kind
    public var lineId: String?
    public var section: String?
    public var text: String?
    /// `.some(nil)` means "place first"; `nil` means "leave where it is" or "append".
    public var afterLineId: String??
    public var supersedes: [String]?
    public var referenceId: String?
}

public struct DraftOperationOutcome: Sendable {
    public var op: String
    public var lineId: String?
    public var accepted: Bool
    /// Written for the model to act on: it says what to do, not just what went wrong.
    public var reason: String?
    public var ratio: Double?
    public var kind: GroundingKind?
    public var unmatchedTokens: [String]?
    public var closestSource: String?
}

private let orderStep: Double = 1000

/// A take that has been sent or thrown away. Nothing may reopen or alter it.
public func isTerminal(_ take: Take) -> Bool {
    take.status == .submitted || take.status == .discarded
}

/// One draft prompt. A session can hold several, so a change of subject does not destroy the last one.
public final class Take: @unchecked Sendable {
    public let id: String
    public var label: String?
    public let createdAt: String
    public var updatedAt: String
    public var status: TakeStatus = .drafting
    public var title: (text: String, origin: String)?
    public var target: String?

    private var linesById: [String: Line] = [:]
    private var contextById: [String: ContextItem] = [:]
    private var contextOrder: [String] = []
    private var pastLines: [Line] = []
    private var sequence = 0

    public init(id: String, createdAt: String, label: String? = nil) {
        self.id = id
        self.createdAt = createdAt
        self.updatedAt = createdAt
        self.label = label
    }

    public func nextLineId() -> String {
        sequence += 1
        return "\(id)-l\(sequence)"
    }

    /// Lines in reading order: section order first, then position within the section.
    public func lines() -> [Line] {
        linesById.values.sorted { lhs, rhs in
            let left = Section.allCases.firstIndex(of: lhs.section) ?? 0
            let right = Section.allCases.firstIndex(of: rhs.section) ?? 0
            if left != right { return left < right }
            if lhs.order != rhs.order { return lhs.order < rhs.order }
            return lhs.id < rhs.id
        }
    }

    public func lines(in section: Section) -> [Line] {
        lines().filter { $0.section == section }
    }

    public func line(_ id: String) -> Line? { linesById[id] }

    /// In the order the speaker brought each reference up, matching the reference implementation.
    /// Sorting by id instead would reorder the rendered context block, since ids come from the host.
    public func context() -> [ContextItem] {
        contextOrder.compactMap { contextById[$0] }
    }

    /// Lines replaced by a correction. Kept so a change of mind can be walked back.
    public func history() -> [Line] { pastLines }

    public func setLine(_ line: Line) {
        if let existing = linesById[line.id] { pastLines.append(existing) }
        linesById[line.id] = line
    }

    @discardableResult
    public func removeLine(_ id: String) -> Bool {
        guard let existing = linesById[id] else { return false }
        pastLines.append(existing)
        linesById[id] = nil
        return true
    }

    public func attach(_ item: ContextItem) {
        if contextById[item.referenceId] == nil { contextOrder.append(item.referenceId) }
        contextById[item.referenceId] = item
    }

    @discardableResult
    public func detach(_ referenceId: String) -> Bool {
        guard contextById.removeValue(forKey: referenceId) != nil else { return false }
        contextOrder.removeAll { $0 == referenceId }
        return true
    }

    public var isEmpty: Bool { linesById.isEmpty && contextById.isEmpty }

    /// Order value that places a new line immediately after `afterId` within its section.
    public func order(in section: Section, after afterId: String??) -> Double {
        let siblings = lines(in: section)
        switch afterId {
        case .none:
            return (siblings.last?.order ?? 0) + orderStep
        case .some(.none):
            return (siblings.first?.order ?? orderStep * 2) - orderStep
        case .some(.some(let anchorId)):
            guard let index = siblings.firstIndex(where: { $0.id == anchorId }) else {
                return (siblings.last?.order ?? 0) + orderStep
            }
            let anchor = siblings[index]
            let next = index + 1 < siblings.count ? siblings[index + 1] : nil
            return next.map { (anchor.order + $0.order) / 2 } ?? anchor.order + orderStep
        }
    }

    /// Whether the take has everything the policy says a sendable prompt needs.
    public func isReady(policy: PolicyConfig) -> Bool {
        policy.readinessRequires.allSatisfy { !lines(in: $0).isEmpty }
    }
}

public struct ApplyContext: Sendable {
    public var checker: GroundingChecker
    public var spans: [SourceSpan]
    public var grounding: GroundingConfig
    /// Resolved references the agent may attach, keyed by reference id.
    public var references: [String: ContextItem]
    public var motifs: [String: Motif]
    public var now: String
}

/// Applies the agent's draft operations, refusing any line that is not made of the speaker's words.
///
/// Rejection is the mechanism that keeps the prompt in their voice: the model proposes text, this
/// function decides whether it survives, and the rejection message tells the model exactly which
/// words it invented so it can put theirs back.
public func applyDraftOperations(
    to take: Take,
    operations: [DraftOperation],
    context: ApplyContext
) -> (accepted: [DraftOperationOutcome], rejected: [DraftOperationOutcome]) {
    var accepted: [DraftOperationOutcome] = []
    var rejected: [DraftOperationOutcome] = []

    func record(_ outcome: DraftOperationOutcome) {
        if outcome.accepted { accepted.append(outcome) } else { rejected.append(outcome) }
    }

    for operation in operations {
        switch operation.op {
        case .setTitle:
            record(applySetTitle(take, operation, context))

        case .upsertLine:
            record(applyUpsert(take, operation, context))

        case .removeLine:
            if let id = operation.lineId, take.removeLine(id) {
                record(DraftOperationOutcome(op: operation.op.rawValue, lineId: id, accepted: true))
            } else {
                record(DraftOperationOutcome(
                    op: operation.op.rawValue,
                    lineId: operation.lineId,
                    accepted: false,
                    reason: "no line \(operation.lineId ?? "(missing line_id)") in this take"
                ))
            }

        case .moveLine:
            if let id = operation.lineId, var line = take.line(id) {
                line.order = take.order(in: line.section, after: operation.afterLineId)
                take.setLine(line)
                record(DraftOperationOutcome(op: operation.op.rawValue, lineId: id, accepted: true))
            } else {
                record(DraftOperationOutcome(
                    op: operation.op.rawValue,
                    lineId: operation.lineId,
                    accepted: false,
                    reason: "no line \(operation.lineId ?? "(missing line_id)") in this take"
                ))
            }

        case .attachContext:
            if let id = operation.referenceId, let reference = context.references[id] {
                take.attach(reference)
                record(DraftOperationOutcome(op: operation.op.rawValue, accepted: true))
            } else {
                record(DraftOperationOutcome(
                    op: operation.op.rawValue,
                    accepted: false,
                    reason: "unknown reference_id \(operation.referenceId ?? "(missing)"); call resolve_reference first and use an id it returned"
                ))
            }

        case .detachContext:
            let detached = operation.referenceId.map { take.detach($0) } ?? false
            record(DraftOperationOutcome(
                op: operation.op.rawValue,
                accepted: detached,
                reason: detached ? nil : "that reference is not attached"
            ))
        }
    }

    if !accepted.isEmpty { take.updatedAt = context.now }
    return (accepted, rejected)
}

private func applySetTitle(_ take: Take, _ operation: DraftOperation, _ context: ApplyContext) -> DraftOperationOutcome {
    let text = RiffText.tidyWhitespace(operation.text ?? "")
    guard !text.isEmpty else {
        return DraftOperationOutcome(op: "set_title", accepted: false, reason: "set_title needs text")
    }
    let result = context.checker.checkTitle(text, against: context.spans)
    guard result.ok else {
        // A title is held to a looser bar than a body line, but it may still only use words they
        // used. Accepting it as `derived` here would make the looser bar no bar at all.
        return DraftOperationOutcome(
            op: "set_title",
            accepted: false,
            reason: result.reason ?? rejectionReason(result.unmatchedTokens, context.spans),
            ratio: result.ratio,
            unmatchedTokens: result.unmatchedTokens
        )
    }

    take.title = (text, "spoken")
    return DraftOperationOutcome(op: "set_title", accepted: true, ratio: result.ratio, kind: result.kind)
}

private func applyUpsert(_ take: Take, _ operation: DraftOperation, _ context: ApplyContext) -> DraftOperationOutcome {
    let text = RiffText.tidyWhitespace(operation.text ?? "")
    guard let raw = operation.section, let section = Section(rawValue: raw) else {
        return DraftOperationOutcome(
            op: "upsert_line",
            accepted: false,
            reason: "section must be one of \(Section.allCases.map(\.rawValue).joined(separator: ", "))"
        )
    }
    guard !text.isEmpty else {
        return DraftOperationOutcome(op: "upsert_line", accepted: false, reason: "upsert_line needs text")
    }

    let mode = context.grounding.sections[section.rawValue] ?? .strict
    let motif = context.motifs.values.first {
        $0.retiredAt == nil && $0.text.lowercased() == text.lowercased()
    }

    var grounding: Line.Grounding
    var sourceUtteranceIds: [String] = []
    var motifId: String?

    if mode == .motifOrStrict, let motif {
        grounding = Line.Grounding(ratio: 1, kind: .motif)
        motifId = motif.id
    } else {
        let result = context.checker.check(text, against: context.spans)
        guard result.ok else {
            return DraftOperationOutcome(
                op: "upsert_line",
                lineId: operation.lineId,
                accepted: false,
                reason: result.reason ?? rejectionReason(result.unmatchedTokens, context.spans),
                ratio: result.ratio,
                unmatchedTokens: result.unmatchedTokens,
                closestSource: closestSource(context.spans, result.sourceUtteranceIds)
            )
        }
        grounding = Line.Grounding(
            ratio: result.ratio,
            kind: result.kind,
            unmatchedTokens: result.unmatchedTokens.isEmpty ? nil : result.unmatchedTokens
        )
        sourceUtteranceIds = result.sourceUtteranceIds
    }

    let existing = operation.lineId.flatMap { take.line($0) }
    if let supplied = operation.lineId, existing == nil {
        // Accepting an unknown id would create a line outside the generated sequence, and the next
        // ordinary insert would reuse that id and silently overwrite this line.
        return DraftOperationOutcome(
            op: "upsert_line",
            lineId: supplied,
            accepted: false,
            reason: "no line \(supplied) in this take; omit line_id to add a new one"
        )
    }
    let id = existing?.id ?? take.nextLineId()
    let order: Double
    if let existing, operation.afterLineId == nil {
        order = existing.order
    } else {
        order = take.order(in: section, after: operation.afterLineId)
    }

    var supersedes: [String] = []
    for target in operation.supersedes ?? [] where target != id && !supersedes.contains(target) {
        supersedes.append(target)
        take.removeLine(target)
    }

    take.setLine(Line(
        id: id,
        section: section,
        text: text,
        order: order,
        sourceUtteranceIds: sourceUtteranceIds,
        motifId: motifId,
        supersedes: supersedes.isEmpty ? nil : supersedes,
        grounding: grounding
    ))

    return DraftOperationOutcome(
        op: "upsert_line",
        lineId: id,
        accepted: true,
        ratio: grounding.ratio,
        kind: grounding.kind
    )
}

private func rejectionReason(_ unmatched: [String], _ spans: [SourceSpan]) -> String {
    if spans.isEmpty { return "nothing has been said yet, so there is nothing to draw on" }
    if unmatched.isEmpty { return "the words are theirs but the order is not; keep their phrasing intact" }
    let shown = unmatched.prefix(6).map { "\"\($0)\"" }.joined(separator: ", ")
    let more = unmatched.count > 6 ? " and \(unmatched.count - 6) more" : ""
    return "they did not say \(shown)\(more); use their words or ask them"
}

private func closestSource(_ spans: [SourceSpan], _ utteranceIds: [String]) -> String? {
    guard !utteranceIds.isEmpty else { return nil }
    let wanted = Set(utteranceIds)
    return spans.first { !$0.utteranceIds.filter(wanted.contains).isEmpty }?.tokens.joined(separator: " ")
}

/// Holds every take in the session and tracks which one is being spoken into.
public final class DraftBook: @unchecked Sendable {
    private var takesById: [String: Take] = [:]
    private var order: [String] = []
    private var sequence = 0
    private let policy: PolicyConfig
    private let clock: @Sendable () -> String

    public private(set) var activeId: String?

    public init(policy: PolicyConfig, now: @escaping @Sendable () -> String = { ISO8601.now() }) {
        self.policy = policy
        self.clock = now
    }

    public func takes() -> [Take] { order.compactMap { takesById[$0] } }

    public func take(_ id: String) -> Take? { takesById[id] }

    /// The take being spoken into, creating the first one on demand.
    public func active() throws -> Take {
        if let activeId, let take = takesById[activeId] { return take }
        return try create()
    }

    @discardableResult
    public func create(label: String? = nil, carryContextFrom source: Take? = nil) throws -> Take {
        let live = takes().filter { $0.status != .discarded && $0.status != .submitted }
        if live.count >= policy.maxTakes {
            if let oldest = live.min(by: { $0.updatedAt < $1.updatedAt }), oldest.isEmpty {
                takesById[oldest.id] = nil
                order.removeAll { $0 == oldest.id }
            } else {
                throw RiffError.tool("too many open takes (limit \(policy.maxTakes)); park or send one first")
            }
        }

        sequence += 1
        let take = Take(id: "t\(sequence)", createdAt: clock(), label: label)
        for item in source?.context() ?? [] { take.attach(item) }
        takesById[take.id] = take
        order.append(take.id)
        activeId = take.id
        return take
    }

    @discardableResult
    public func switchTo(_ id: String) throws -> Take? {
        guard let take = takesById[id] else { return nil }
        // Switching to a finished take would make it the target of the next thing spoken.
        guard !isTerminal(take) else {
            throw RiffError.tool("take \"\(id)\" was already \(take.status.rawValue); it cannot be reopened")
        }
        if let activeId, activeId != id, let previous = takesById[activeId], previous.status == .drafting {
            previous.status = .parked
        }
        if take.status == .parked { take.status = .drafting }
        activeId = id
        return take
    }

    @discardableResult
    public func park(_ id: String) throws -> Take? {
        guard let take = takesById[id] else { return nil }
        // Parking a finished take would move it out of a terminal state, and switching back would
        // then promote it to drafting — which is how every guard downstream gets bypassed.
        guard !isTerminal(take) else {
            throw RiffError.tool("take \"\(id)\" was already \(take.status.rawValue); it cannot be parked")
        }
        take.status = .parked
        if activeId == id {
            activeId = takes().first { $0.id != id && $0.status == .drafting }?.id
        }
        return take
    }

    @discardableResult
    public func discard(_ id: String) throws -> Bool {
        guard let take = takesById[id] else { return false }
        guard take.status != .submitted else {
            throw RiffError.tool("take \"\(id)\" was already submitted")
        }
        take.status = .discarded
        if activeId == id { activeId = nil }
        return true
    }

    public func clearActive() { activeId = nil }
}

File: packages/riff-core/src/draft.ts
import type {
  ContextItem,
  GroundingConfig,
  Line,
  Motif,
  PolicyConfig,
  Section,
  TakeStatus,
} from "./types.ts";
import { SECTIONS, isSection } from "./types.ts";
import type { GroundingChecker, SourceSpan } from "./grounding.ts";
import { tidyWhitespace } from "./text.ts";

export interface DraftOperation {
  op: "set_title" | "upsert_line" | "remove_line" | "move_line" | "attach_context" | "detach_context";
  line_id?: string;
  section?: string;
  text?: string;
  after_line_id?: string | null;
  supersedes?: string[];
  reference_id?: string;
}

export interface DraftOperationOutcome {
  op: DraftOperation["op"];
  lineId?: string;
  status: "accepted" | "rejected";
  /** Written for the model to act on: it says what to do, not just what went wrong. */
  reason?: string;
  ratio?: number;
  kind?: Line["grounding"]["kind"];
  unmatchedTokens?: string[];
  closestSource?: string;
}

const ORDER_STEP = 1000;

/** A take that has been sent or thrown away. Nothing may reopen or alter it. */
export function isTerminal(take: Take): boolean {
  return take.status === "submitted" || take.status === "discarded";
}

/** One draft prompt. A session can hold several, so a change of subject does not destroy the last one. */
export class Take {
  readonly id: string;
  label?: string;
  readonly createdAt: string;
  updatedAt: string;
  status: TakeStatus = "drafting";
  title: { text: string; origin: "derived" | "spoken" } | null = null;
  target?: string;

  #lines = new Map<string, Line>();
  #context = new Map<string, ContextItem>();
  #history: Line[] = [];
  #sequence = 0;

  constructor(id: string, createdAt: string, label?: string) {
    this.id = id;
    this.createdAt = createdAt;
    this.updatedAt = createdAt;
    if (label !== undefined) this.label = label;
  }

  nextLineId(): string {
    return `${this.id}-l${++this.#sequence}`;
  }

  /** Lines in reading order: section order first, then position within the section. */
  lines(): Line[] {
    return [...this.#lines.values()].sort(
      (a, b) => SECTIONS.indexOf(a.section) - SECTIONS.indexOf(b.section) || a.order - b.order,
    );
  }

  linesIn(section: Section): Line[] {
    return this.lines().filter((line) => line.section === section);
  }

  line(id: string): Line | undefined {
    return this.#lines.get(id);
  }

  context(): ContextItem[] {
    return [...this.#context.values()];
  }

  /** Lines replaced by a correction. Kept so a change of mind can be walked back. */
  history(): readonly Line[] {
    return this.#history;
  }

  setLine(line: Line): void {
    const existing = this.#lines.get(line.id);
    if (existing) this.#history.push({ ...existing });
    this.#lines.set(line.id, line);
  }

  removeLine(id: string): boolean {
    const existing = this.#lines.get(id);
    if (!existing) return false;
    this.#history.push({ ...existing });
    this.#lines.delete(id);
    return true;
  }

  attachContext(item: ContextItem): void {
    this.#context.set(item.referenceId, item);
  }

  detachContext(referenceId: string): boolean {
    return this.#context.delete(referenceId);
  }

  isEmpty(): boolean {
    return this.#lines.size === 0 && this.#context.size === 0;
  }

  /** Order value that places a new line immediately after `afterId` within its section. */
  orderAfter(section: Section, afterId: string | null | undefined): number {
    const siblings = this.linesIn(section);
    if (afterId === null) {
      const first = siblings[0];
      return first ? first.order - ORDER_STEP : ORDER_STEP;
    }
    if (afterId === undefined) {
      const last = siblings[siblings.length - 1];
      return last ? last.order + ORDER_STEP : ORDER_STEP;
    }
    const index = siblings.findIndex((line) => line.id === afterId);
    if (index === -1) {
      const last = siblings[siblings.length - 1];
      return last ? last.order + ORDER_STEP : ORDER_STEP;
    }
    const anchor = siblings[index];
    const next = siblings[index + 1];
    if (!anchor) return ORDER_STEP;
    return next ? (anchor.order + next.order) / 2 : anchor.order + ORDER_STEP;
  }
}

export interface ApplyContext {
  checker: GroundingChecker;
  spans: readonly SourceSpan[];
  grounding: GroundingConfig;
  /** Resolved references the agent may attach, keyed by reference id. */
  references: Map<string, ContextItem>;
  motifs: Map<string, Motif>;
  now: () => string;
}

export interface ApplyResult {
  accepted: DraftOperationOutcome[];
  rejected: DraftOperationOutcome[];
}

/**
 * Applies the agent's draft operations, refusing any line that is not made of the speaker's words.
 *
 * Rejection is the mechanism that keeps the prompt in their voice: the model proposes text, this
 * function decides whether it survives, and the rejection message tells the model exactly which
 * words it invented so it can put theirs back.
 */
export function applyDraftOperations(
  take: Take,
  operations: readonly DraftOperation[],
  ctx: ApplyContext,
): ApplyResult {
  const accepted: DraftOperationOutcome[] = [];
  const rejected: DraftOperationOutcome[] = [];
  const record = (outcome: DraftOperationOutcome) =>
    (outcome.status === "accepted" ? accepted : rejected).push(outcome);

  for (const operation of operations) {
    switch (operation.op) {
      case "set_title":
        record(applySetTitle(take, operation, ctx));
        break;
      case "upsert_line":
        record(applyUpsertLine(take, operation, ctx));
        break;
      case "remove_line":
        record(
          take.removeLine(operation.line_id ?? "")
            ? { op: operation.op, lineId: operation.line_id, status: "accepted" }
            : {
                op: operation.op,
                lineId: operation.line_id,
                status: "rejected",
                reason: `no line ${operation.line_id ?? "(missing line_id)"} in this take`,
              },
        );
        break;
      case "move_line":
        record(applyMoveLine(take, operation));
        break;
      case "attach_context": {
        const reference = operation.reference_id ? ctx.references.get(operation.reference_id) : undefined;
        record(
          reference
            ? (take.attachContext(reference), { op: operation.op, status: "accepted" })
            : {
                op: operation.op,
                status: "rejected",
                reason: `unknown reference_id ${operation.reference_id ?? "(missing)"}; call resolve_reference first and use an id it returned`,
              },
        );
        break;
      }
      case "detach_context":
        record(
          take.detachContext(operation.reference_id ?? "")
            ? { op: operation.op, status: "accepted" }
            : { op: operation.op, status: "rejected", reason: "that reference is not attached" },
        );
        break;
      default:
        record({
          op: operation.op,
          status: "rejected",
          reason: `unknown operation "${String(operation.op)}"`,
        });
    }
  }

  if (accepted.length > 0) take.updatedAt = ctx.now();
  return { accepted, rejected };
}

function applySetTitle(take: Take, operation: DraftOperation, ctx: ApplyContext): DraftOperationOutcome {
  const text = tidyWhitespace(operation.text ?? "");
  if (!text) return { op: "set_title", status: "rejected", reason: "set_title needs text" };

  const result = ctx.checker.checkTitle(text, ctx.spans);
  if (!result.ok) {
    // A title is held to a looser bar than a body line, but it may still only use words they used.
    // Accepting it as `derived` here would make the looser bar no bar at all.
    return {
      op: "set_title",
      status: "rejected",
      reason: result.reason ?? rejectionReason(result.unmatchedTokens, ctx.spans),
      ratio: result.ratio,
      unmatchedTokens: result.unmatchedTokens,
    };
  }

  take.title = { text, origin: "spoken" };
  return { op: "set_title", status: "accepted", ratio: result.ratio, kind: result.kind };
}

function applyUpsertLine(take: Take, operation: DraftOperation, ctx: ApplyContext): DraftOperationOutcome {
  const text = tidyWhitespace(operation.text ?? "");
  const section = operation.section;

  if (!isSection(section)) {
    return {
      op: "upsert_line",
      status: "rejected",
      reason: `section must be one of ${SECTIONS.join(", ")}`,
    };
  }
  if (!text) return { op: "upsert_line", status: "rejected", reason: "upsert_line needs text" };

  const mode = ctx.grounding.sections[section] ?? "strict";
  const motif = findMotif(text, ctx.motifs);

  let grounding: Line["grounding"];
  let sourceUtteranceIds: string[] = [];
  let motifId: string | undefined;

  if (mode === "motif-or-strict" && motif) {
    grounding = { ratio: 1, kind: "motif" };
    motifId = motif.id;
  } else {
    const result = ctx.checker.check(text, ctx.spans);
    if (!result.ok) {
      return {
        op: "upsert_line",
        lineId: operation.line_id,
        status: "rejected",
        reason: result.reason ?? rejectionReason(result.unmatchedTokens, ctx.spans),
        ratio: result.ratio,
        unmatchedTokens: result.unmatchedTokens,
        closestSource: closestSourceText(ctx.spans, result.sourceUtteranceIds),
      };
    }
    grounding = {
      ratio: result.ratio,
      kind: result.kind,
      ...(result.unmatchedTokens.length > 0 ? { unmatchedTokens: result.unmatchedTokens } : {}),
    };
    sourceUtteranceIds = result.sourceUtteranceIds;
  }

  const existing = operation.line_id ? take.line(operation.line_id) : undefined;
  if (operation.line_id && !existing) {
    // Accepting an unknown id would create a line outside the generated sequence, and the next
    // ordinary insert would reuse that id and silently overwrite this line.
    return {
      op: "upsert_line",
      lineId: operation.line_id,
      status: "rejected",
      reason: `no line ${operation.line_id} in this take; omit line_id to add a new one`,
    };
  }
  const id = existing?.id ?? take.nextLineId();
  const order =
    existing && operation.after_line_id === undefined
      ? existing.order
      : take.orderAfter(section, operation.after_line_id);

  const supersedes = [...new Set(operation.supersedes ?? [])].filter((target) => target !== id);
  for (const target of supersedes) take.removeLine(target);

  take.setLine({
    id,
    section,
    text,
    order,
    sourceUtteranceIds,
    ...(motifId ? { motifId } : {}),
    ...(supersedes.length > 0 ? { supersedes } : {}),
    grounding,
  });

  return { op: "upsert_line", lineId: id, status: "accepted", ratio: grounding.ratio, kind: grounding.kind };
}

function applyMoveLine(take: Take, operation: DraftOperation): DraftOperationOutcome {
  const line = operation.line_id ? take.line(operation.line_id) : undefined;
  if (!line) {
    return {
      op: "move_line",
      lineId: operation.line_id,
      status: "rejected",
      reason: `no line ${operation.line_id ?? "(missing line_id)"} in this take`,
    };
  }
  take.setLine({ ...line, order: take.orderAfter(line.section, operation.after_line_id) });
  return { op: "move_line", lineId: line.id, status: "accepted" };
}

function findMotif(text: string, motifs: Map<string, Motif>): Motif | undefined {
  const normalized = text.toLowerCase();
  for (const motif of motifs.values()) {
    if (motif.retiredAt) continue;
    if (motif.text.toLowerCase() === normalized) return motif;
  }
  return undefined;
}

function rejectionReason(unmatched: string[], spans: readonly SourceSpan[]): string {
  if (spans.length === 0) return "nothing has been said yet, so there is nothing to draw on";
  if (unmatched.length === 0) return "the words are theirs but the order is not; keep their phrasing intact";
  const shown = unmatched.slice(0, 6).map((token) => `"${token}"`).join(", ");
  const more = unmatched.length > 6 ? ` and ${unmatched.length - 6} more` : "";
  return `they did not say ${shown}${more}; use their words or ask them`;
}

function closestSourceText(spans: readonly SourceSpan[], utteranceIds: string[]): string | undefined {
  if (utteranceIds.length === 0) return undefined;
  const wanted = new Set(utteranceIds);
  const match = spans.find((span) => span.utteranceIds.some((id) => wanted.has(id)));
  return match?.tokens.join(" ");
}

/** Whether the take has everything the policy says a sendable prompt needs. */
export function isReady(take: Take, policy: PolicyConfig): boolean {
  return policy.readinessRequires.every((section) => take.linesIn(section).length > 0);
}

/** Holds every take in the session and tracks which one is being spoken into. */
export class DraftBook {
  #takes = new Map<string, Take>();
  #activeId: string | null = null;
  #sequence = 0;
  readonly #policy: PolicyConfig;
  readonly #now: () => string;

  constructor(policy: PolicyConfig, now: () => string = () => new Date().toISOString()) {
    this.#policy = policy;
    this.#now = now;
  }

  get activeId(): string | null {
    return this.#activeId;
  }

  takes(): Take[] {
    return [...this.#takes.values()];
  }

  get(id: string): Take | undefined {
    return this.#takes.get(id);
  }

  /** The take being spoken into, creating the first one on demand. */
  active(): Take {
    if (this.#activeId) {
      const take = this.#takes.get(this.#activeId);
      if (take) return take;
    }
    return this.create();
  }

  create(label?: string, carryContextFrom?: Take): Take {
    const live = this.takes().filter((take) => take.status !== "discarded" && take.status !== "submitted");
    if (live.length >= this.#policy.maxTakes) {
      const oldest = live.sort((a, b) => a.updatedAt.localeCompare(b.updatedAt))[0];
      if (oldest && oldest.isEmpty()) this.#takes.delete(oldest.id);
      else throw new Error(`too many open takes (limit ${this.#policy.maxTakes}); park or send one first`);
    }

    const take = new Take(`t${++this.#sequence}`, this.#now(), label);
    if (carryContextFrom) for (const item of carryContextFrom.context()) take.attachContext(item);
    this.#takes.set(take.id, take);
    this.#activeId = take.id;
    return take;
  }

  switchTo(id: string): Take | undefined {
    const take = this.#takes.get(id);
    if (!take) return undefined;
    // Switching to a finished take would make it the target of the next thing spoken.
    if (isTerminal(take)) throw new Error(`take "${id}" was already ${take.status}; it cannot be reopened`);
    const previous = this.#activeId ? this.#takes.get(this.#activeId) : undefined;
    if (previous && previous.id !== id && previous.status === "drafting") previous.status = "parked";
    take.status = take.status === "parked" ? "drafting" : take.status;
    this.#activeId = id;
    return take;
  }

  park(id: string): Take | undefined {
    const take = this.#takes.get(id);
    if (!take) return undefined;
    // Parking a finished take would move it out of a terminal state, and switching back would then
    // promote it to drafting — which is how every guard downstream gets bypassed.
    if (isTerminal(take)) throw new Error(`take "${id}" was already ${take.status}; it cannot be parked`);
    take.status = "parked";
    if (this.#activeId === id) {
      const next = this.takes().find((candidate) => candidate.id !== id && candidate.status === "drafting");
      this.#activeId = next?.id ?? null;
    }
    return take;
  }

  /** Leaves no take active, so the next line spoken starts a fresh one. */
  clearActive(): void {
    this.#activeId = null;
  }

  discard(id: string): boolean {
    const take = this.#takes.get(id);
    if (!take) return false;
    if (take.status === "submitted") throw new Error(`take "${id}" was already submitted`);
    take.status = "discarded";
    if (this.#activeId === id) this.#activeId = null;
    return true;
  }
}
